jetstream/probeinfo.py

```diff
@@ -0,0 +1,31 @@
+from typing import Set
+
+from glom import glom, Flatten, Iter, Path, T
+import requests
+
+
+class _DesktopProbeInfo:
+    PROBE_INFO_URL = "https://probeinfo.telemetry.mozilla.org/firefox/all/main/all_probes"
+
+    @property
+    def data(self):
+        if data := getattr(self, "_data", None):
+            return data
+        self._data = requests.get(self.PROBE_INFO_URL).json()
+        return self._data
+
+    def processes_for_scalar(self, name) -> Set[str]:
+        scalar_key = f"scalar/{name}"
+        return glom(
+            self.data,
+            (
+                Path(scalar_key, "history"),
+                T.values(),  # all release channels
+                Flatten(),  # all history
+                Iter("details.record_in_processes").flatten(),
```

Comment:
Can you tell me more about completely inaccurate?

I am not surprised if the probe scraper includes processes that do not actually or only rarely report the probe; engineers sometimes list more processes than necessary in the probe definitions. It's no problem if we're looking in columns that exist but happen to contain no data. Here, we just want to make sure we have coverage of all places where data for a probe could have appeared.

If the probe scraper is not reporting a complete set of processes for a probe, that is a serious bug with the probe scraper.

If the probe scraper disagrees with the data pipeline, it would be useful to understand why.

Reply:
I think the most serious issue is when the probe info service reports that a process contains a probe but that probe isn't in the list for that process at all in the table. For example, `scalar/telemetry.discarded.keyed_scalar_actions` in probe info service lists `all_childs` in `record_in_processes`. 

However if you select `payload.processes.parent.scalars.telemetry_discarded_keyed_scalar_actions`, it gives an error because `telemetry_discarded_keyed_scalar_actions` is not in `payload.processes.parent.scalars`  but it works for `content` and `gpu` processes.

I think if you want a deeper understanding why the probe info disagrees with the data, chutten is the person to talk to as he is the one who wrote the `record_in_processes` bit and he is the one who told us it's inaccurate.

If I were to explain at a high level why, it sounded like a design decision that happened years ago to list as many processes as possible in `record_in_processes` although the data may not be there at all.

Reply:
Hm -- that is the case I'm concerned about, but that example sounds consistent to me, because main/parent isn't a child process.

I'm not handling "all_childs" though, eep -- it looks like a list of things I need to worry about is here: https://searchfox.org/mozilla-central/source/toolkit/components/telemetry/build_scripts/mozparsers/shared_telemetry_utils.py#17-26

Ultimately, what matters here is what mozilla-schema-generator is doing, since that's what produces the BigQuery schema. And it looks like mozilla-schema-generator is:
a) using the probeinfo service: https://github.com/mozilla/mozilla-schema-generator/blob/master/mozilla_schema_generator/generic_ping.py#L21, and
b) using recorded_in_processes the same way this code does: https://github.com/mozilla/mozilla-schema-generator/blob/master/mozilla_schema_generator/probes.py#L101

So I think I'm convinced this is actually the right thing to do (subject to handling the special process aliases correctly).

@chutten, do you have any concerns about relying on recorded_in_processes from the probe scraper to understand which columns exist in BigQuery (whether they're populated or not)?

Reply:
I guess maybe another way to look at this is whether we have a preference for generating column names from the metadata or filtering column names from the schema. We can retrieve the schema from BigQuery itself, which cuts to the heart of the source-of-truth question. That sounds probably less error-prone than trying to match the logic of every other part of the pipeline.

Reply:
> Hm -- that is the case I'm concerned about, but that example sounds consistent to me, because main/parent isn't a child process.

Ok I misinterpreted `all_childs` to mean "all processes".

Yes, the case in Glam was mostly focused on listed processes having no data. Thanks for the links, that does in fact look safe to do then.

Reply:
Ideally we'd be able to query BQ directly for that info instead of duplicating the logic and hoping we don't drift too quickly, but I'm not planning any changes other than digging the whole thing out after we migrate everyone to Glean so it should be fine.

Reply:
> We can retrieve the schema from BigQuery itself, which cuts to the heart of the source-of-truth question. That sounds probably less error-prone than trying to match the logic of every other part of the pipeline.

This sounds a lot safer to me as well.

Reply:
Done; removed all this code ✂️ Thanks for the discussion!